src/dataFiles.ts

```diff
@@ -39,65 +36,104 @@ export async function initializeDataDirs(): Promise<void> {
+			}, []);
+			return indexDataPaths(paths);
+		}),
+	);
 }
 
-async function indexDataDir(
-	dataDir: string,
-	prevWatcher: FSWatcher | null,
-): Promise<void> {
-	const { maxDataDepth, seasonFromEpisodes } = getRuntimeConfig();
-	if (prevWatcher) return;
-	logger.verbose(`Indexing ${dataDir} due to recent changes...`);
-	watchers.set(dataDir, startWatcher(dataDir));
+/**
+ * Adds the data and ensemble entries to the database.
+ * @param paths The paths to index.
+ */
+async function indexDataPaths(paths: string[]): Promise<void> {
+	const { seasonFromEpisodes } = getRuntimeConfig();
 	const memoizedPaths = new Map<string, string[]>();
 	const memoizedLengths = new Map<string, number>();
-	for (const dirent of readdirSync(dataDir)) {
-		const entry = join(dataDir, dirent);
-		const paths = findPotentialNestedRoots(entry, maxDataDepth);
-		const dataRows: DataEntry[] = [];
-		const ensembleRows: EnsembleEntry[] = [];
-		for (const path of paths) {
-			const files = getFilesFromDataRoot(
+	const dataRows: DataEntry[] = [];
+	const ensembleRows: EnsembleEntry[] = [];
+	for (const path of paths) {
+		const files = getFilesFromDataRoot(
+			path,
+			memoizedPaths,
+			memoizedLengths,
+		);
+		const title = parseTitle(basename(path), files, path);
+		if (!title) continue;
+		dataRows.push({ title, path });
+		if (seasonFromEpisodes) {
+			const ensembleEntry = await indexEnsembleDataEntry(
+				title,
 				path,
-				memoizedPaths,
-				memoizedLengths,
+				files,
 			);
-			const title = parseTitle(basename(path), files, path);
-			if (!title) continue;
-			dataRows.push({ title, path });
-			if (seasonFromEpisodes) {
-				const ensembleEntry = await indexEnsembleDataEntry(
-					title,
-					path,
-					files,
-				);
-				if (ensembleEntry) ensembleRows.push(ensembleEntry);
-			}
-		}
-		const batchSize = 100;
-		for (let i = 0; i < dataRows.length; i += batchSize) {
-			const batch = dataRows.slice(i, i + batchSize);
-			if (!batch.length) break;
-			await memDB("data").insert(batch).onConflict("path").ignore();
-		}
-		for (let i = 0; i < ensembleRows.length; i += batchSize) {
-			const batch = ensembleRows.slice(i, i + batchSize);
-			if (!batch.length) break;
-			await memDB("ensemble").insert(batch).onConflict("path").ignore();
+			if (ensembleEntry) ensembleRows.push(ensembleEntry);
 		}
 	}
-}
-
-export async function indexDataDirs(): Promise<void> {
-	await Promise.all(
-		Array.from(watchers.entries()).map(([dataDir, prevWatcher]) =>
-			indexDataDir(dataDir, prevWatcher),
-		),
-	);
+	const batchSize = 100;
+	for (let i = 0; i < dataRows.length; i += batchSize) {
+		const batch = dataRows.slice(i, i + batchSize);
+		if (!batch.length) break;
+		await memDB("data").insert(batch).onConflict("path").merge();
+	}
+	for (let i = 0; i < ensembleRows.length; i += batchSize) {
+		const batch = ensembleRows.slice(i, i + batchSize);
+		if (!batch.length) break;
+		await memDB("ensemble").insert(batch).onConflict("path").merge();
```

Comment:
knex has a batch insert util that could help here

Reply:
`memDB.batchInsert("data", dataRows, batchSize)` seems like it would work but it doesn't support `.onConflict()`.